File: preprocess/farthest_point_sampling_fast.m
function sampled_idx = farthest_point_sampling_fast(point_cloud, sample_num)
% farthest point sampling
% point_cloud: Nx3


pc_num = size(point_cloud,1);

if pc_num <= sample_num
    sampled_idx = [1:pc_num]';
        % randi([min,max],m,n)生成在min~max范围内的随机m*n矩阵
    sampled_idx = [sampled_idx; randi([1,pc_num],sample_num-pc_num,1)];
else
    sampled_idx = zeros(sample_num,1);
    sampled_idx(1) = randi([1,pc_num]);
    
    cur_sample = repmat(point_cloud(sampled_idx(1),:),pc_num,1);
    diff = point_cloud - cur_sample;
        %sum(A,n)n=1 行求和，n=2列求和
    min_dist = sum(diff.*diff, 2);

    for cur_sample_idx = 2:sample_num
        %% find the farthest point
        [~, sampled_idx(cur_sample_idx)] = max(min_dist);
        
        if cur_sample_idx < sample_num
            % update min_dist
            valid_idx = (min_dist>1e-8);    % 根据条件返回逻辑值
            diff = point_cloud(valid_idx,:) - repmat(point_cloud(sampled_idx(cur_sample_idx),:),sum(valid_idx),1);
            min_dist(valid_idx,:) = min(min_dist(valid_idx,:), sum(diff.*diff, 2));
        end
    end
end
    %A=unique(a) 返回a中不相同元素的集合
sampled_idx = unique(sampled_idx);

File: pre/farthest_point_sampling_fast.py
'''
farthest point sampling
point_cloud: Nx3
'''

import numpy as np


class FPSF(object):
    def __init__(self, point_cloud, sample_num):
        self.point_cloud = point_cloud
        self.sample_num = sample_num
        self.pc_num = point_cloud.shape[0]

    def fpsf(self):

        if self.pc_num <= self.sample_num:
            sampled_idx = np.array(
                range(self.pc_num), dtype=np.int32).transpose()
            rand_m = np.random.randint(
                1, self.pc_num, size=(self.sample_num-self.pc_num, 1))
            sampled_idx = np.append(sampled_idx, rand_m, axis=0)
        else:
            sampled_idx = np.zeros(self.sample_num, 1)
            sampled_idx[1] = np.random.randint(1, self.sample_num)

            diff_n = self.point_cloud - sampled_idx
            min_dist = np.sum(np.dot(diff_n, diff_n), axis=1)

            # 可能需要修改
            for i in [i for i in range(self.sample_num)].pop(0):
                sampled_idx[i] = np.max(min_dist)
                if i < self.sample_num:
                    valid_idx = [min_dist > np.exp(-8)]
                    diff_n = self.point_cloud(
                        valid_idx) - self.point_cloud(sampled_idx)
                    for i in valid_idx:
                        min_dist[valid_idx[i], :] = np.min(
                            valid_idx, axis=np.sum(np.dot(diff_n, diff_n), axis=1))

        sampled_idx = np.unique(sampled_idx)

        return sampled_idx
